File: code/python/剑指offer04.py
class Solution:
    def findNumberIn2DArray(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix: return False
        m = len(matrix)
        n = len(matrix[0])
        i = 0
        j = n-1
        while i>=0 and i<m and j>=0 and j<n:
            if matrix[i][j]>target:
                j-=1
            elif matrix[i][j]<target:
                i+=1
            else:
                return True

        return False

File: code/java/剑指offer04.java
class Solution {
    public boolean findNumberIn2DArray(int[][] matrix, int target) {
        if(matrix==null||matrix.length==0) return false;
        int m = matrix.length;
        int n = matrix[0].length;
        int i = 0;
        int j = n-1;
        while(i>=0&&i<m&&j>=0&&j<n){
            if(matrix[i][j]>target) j--;
            else if(matrix[i][j]<target) i++;
            else if(matrix[i][j]==target) return true;
        }
        return false;
    }
}